Consider mapped upstream in TriggerRuleDep

uranusjr opened the pull request:
This is needed to ensure all mapped upstream task instances are depended on by a downstream, instead of just one of them signifying the entire task as "done".

~Does not work yet. This depends on correct task-unmapping behaviour implemented in #22396. I also want to add some more asserts to make sure the downstream can pull all the values from the mapped upstream, but that depends on #22609.~ Ready!

uranusjr committed:
Consider mapped upstream in TriggerRuleDep

This is needed to ensure all mapped upstream task instances are depended
on by a downstream, instead of just one of them signifying the entire
task as "done".

ashb commented on airflow/models/mappedoperator.py:
```diff
@@ -542,7 +542,7 @@ def expand_mapped_task(self, run_id: str, *, session: Session) -> Sequence["Task
             state = unmapped_ti.state
             self.log.debug("Updated in place to become %s", unmapped_ti)
-            ret.append(unmapped_ti)
+            ret.append(session.merge(unmapped_ti))
```
This shouldn't be required as the `unmapped_ti` is still attached to `session`

uranusjr committed:
Ensure downstream "joins" a mapped upstream

We catually missed quite several things for this to happen! Most namely
the pre-execution XCom clearing behavior; that's surprising. But things
should work now.

uranusjr commented on airflow/models/taskinstance.py:
```diff
@@ -819,17 +819,25 @@ def refresh_from_task(self, task: "Operator", pool_override=None):
-    def clear_xcom_data(self, session=NEW_SESSION):
-        """
-        Clears all XCom data from the database for the task instance
+    def clear_xcom_data(self, session: Session = NEW_SESSION):
+        """Clear all XCom data from the database for the task instance.
+
+        If the task is unmapped, all XComs matching this task ID in the same DAG
+        run are removed. If the task is mapped, only the one with matching map
+        index is removed.
 
         :param session: SQLAlchemy ORM Session
         """
         self.log.debug("Clearing XCom data")
+        if self.map_index < 0:
+            map_index: Optional[int] = None
+        else:
+            map_index = self.map_index
         XCom.clear(
             dag_id=self.dag_id,
             task_id=self.task_id,
             run_id=self.run_id,
+            map_index=map_index,
             session=session,
         )
```
Somewhat surprising we didn’t catch this earlier!

dstandish commented on airflow/models/taskinstance.py:
```diff
@@ -2339,7 +2350,15 @@ def xcom_pull(
+            def iter_xcom_values(query):
+                # The session passed to xcom_pull() may die before this is
+                # iterated through, so we need to bind to a new session.
+                with create_session() as session:
```
why create a new one?

uranusjr replied:
Because `session` in the outer scope can die before we iterate through this, e.g.

```python
@task
def double(v):
    return v * 2

@task
def consume(v):
    resolved = list(v)  # The original session in xcom_pull is dead by now.

consume(v=double.expand(v=[1, 2]))
```

dstandish replied:
I'll take your word for it.  But that example seems quite odd... because xcom_pull isn't actually going to be invoked in what you've written, because no tasks have been run, so there's nothing to pull ...

uranusjr replied:
The return value from `double` tasks are pushed into XCom. This is similar to

```python
@task
def double(v):
    return v * 2

@task
def consume(v):
    print(v)

consume(v=double(v=2))
```

where `double` would push `4` to XCom.

dstandish replied:
yeah i understand that, what i meant was, the example is just dag code, so the data movement is not happening in what you've written, right?  it only happens once you run the tasks.  i presume that to understand what's going with the sessions i need to understand more details about where / when / how this method is actually invoked. but i'll leave that for another day.

dstandish commented on airflow/models/taskinstance.py:
```diff
@@ -819,17 +819,25 @@ def refresh_from_task(self, task: "Operator", pool_override=None):
         """
         self.log.debug("Clearing XCom data")
+        if self.map_index < 0:
+            map_index: Optional[int] = None
```
why convert -1 to None?

uranusjr replied:
This is for clearing stale XComs; if the task instance is unmapped, there shouldn’t be mapped XComs bound to the task, so we use None to clear them all just in case.

ashb commented on airflow/models/taskinstance.py:
```diff
@@ -2339,7 +2350,15 @@ def xcom_pull(
+                # iterated through, so we need to bind to a new session.
+                with create_session() as session:
+                    for r in query.with_session(session):
+                        yield XCom.deserialize_value(r)
```
We shouldn't use `create_session` here as that has the side effect of doing an expunge_all when this block exits.

We just want a `session = settings.Session()`

```suggestion
                for r in query.with_session(settings.Session()):
                    yield XCom.deserialize_value(r)
```

dstandish commented on airflow/models/taskinstance.py:
```diff
@@ -819,17 +819,25 @@ def refresh_from_task(self, task: "Operator", pool_override=None):
-    def clear_xcom_data(self, session=NEW_SESSION):
-        """
-        Clears all XCom data from the database for the task instance
+    def clear_xcom_data(self, session: Session = NEW_SESSION):
```
this reminds me... thinking about custom xcom backends ... there should be a probably be a "clear data" hook that can be implemented in an xcom backend to purge the external data without having to override this method... of course that's another PR

uranusjr replied:
Is `XCom.clear()` not enough?

ashb commented on airflow/ti_deps/deps/trigger_rule_dep.py:
```diff
@@ -83,6 +87,31 @@ def _get_dep_statuses(self, ti, session, dep_context: DepContext):
+
+        # Optimization: Don't need to hit the database if no upstreams are mapped.
+        upstream_task_ids = ti.task.upstream_task_ids
+        if ti.task.dag and not any(ti.task.dag.get_task(tid).is_mapped for tid in upstream_task_ids):
```
```suggestion
        if ti.task.has_dag() and not any(node.is_mapped for node in ti.task.upstream_list):
```

uranusjr replied:
This doesn’t work well with Mypy because `upstream_list` returns a list of DAGNode, not AbstractOperator.
